pkg/datasource/csvsource/stream.go

```diff
@@ -0,0 +1,158 @@
+package csvsource
+
+import (
+	"encoding/csv"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+//go:generate callbackgen -type Stream
+type Stream struct {
+	types.StandardStream
+	config                    *CsvStreamConfig
+	converter                 ICSVTickConverter
+	marketTradeEventCallbacks []func(e []CsvTick)
+	kLineEventCallbacks       []func(e []types.KLine)
+	orderEventCallbacks       []func(e []types.Order)
+	tradeEventCallbacks       []func(e []types.Trade)
+}
+
+type CsvStreamConfig struct {
+	Interval     types.Interval     `json:"interval"`
+	RateLimit    time.Duration      `json:"rateLimit"`
+	StrategyID   string             `json:"strategyID"`
+	CsvPath      string             `json:"csvPath"`
+	Exchange     types.ExchangeName `json:"exchange"`
+	Symbol       string             `json:"symbol"`
+	BaseCoin     string             `json:"baseCoin"`
+	QuoteCoin    string             `json:"quoteCoin"`
+	TakerFeeRate fixedpoint.Value   `json:"takerFeeRate"`
+	MakerFeeRate fixedpoint.Value   `json:"makerFeeRate"`
+}
+
+func NewStream(cfg *CsvStreamConfig) *Stream {
+	stream := &Stream{
+		StandardStream: types.NewStandardStream(),
+		converter:      NewCSVTickConverter(),
+		config:         cfg,
+	}
+
+	// stream.SetParser(stream.simulateEvents)
+	// stream.SetDispatcher(stream.dispatchEvent)
+	// stream.OnMarketTradeEvent(stream.handleMarketTradeEvent)
+	// stream.OnKLineEvent(stream.handleKLineEvent)
+	// stream.OnOrderEvent(stream.handleOrderEvent)
+	// stream.OnTradeEvent(stream.handleTradeEvent)
+	return stream
+}
+
+func (s *Stream) Simulate() error {
+	var i int
+	// iterate equity series at csv path and stream
+	err := filepath.WalkDir(s.config.CsvPath, func(path string, d fs.DirEntry, err error) error {
```

Comment:
why not use the backtest engine to load the csv?

Reply:
Yes I must admit that I am still in the process of figuring things out so I am a bit slow with getting it right. I saw the cmd/backtest.go and how you assemble a backtest. I really appreciate any help and feedback on this.